packages/plugin-node-modules/sources/NodeModulesLinker.ts

```diff
@@ -661,39 +663,53 @@ async function persistNodeModules(preinstallState: InstallState, installState: N
 
 
   // Delete locations that no longer exist
-  const deleteList: PortablePath[] = [];
+  const deleteList = new Set<PortablePath>();
+  // Delete locations of inner node_modules
+  const innerDeleteList = new Set<PortablePath>();
   for (const {locations} of preinstallState.locatorMap.values()) {
     for (const location of locations) {
       const {locationRoot, segments} = parseLocation(location, {
         skipPrefix: project.cwd,
       });
 
+      let prevNode = prevLocationTree.get(locationRoot);
       let node = locationTree.get(locationRoot);
       let curLocation = locationRoot;
-      if (!node) {
-        deleteList.push(curLocation);
-      } else {
+      if (node) {
         for (const segment of segments) {
           // '.' segment exists only for top-level locator, skip it
           if (segment === '.')
             continue;
 
           curLocation = ppath.join(curLocation, segment);
           node = node.children.get(segment);
+          if (prevNode)
+            prevNode = prevNode.children.get(segment);
+
           if (!node) {
-            deleteList.push(curLocation);
+            deleteList.add(curLocation);
+            // If previous install had inner node_modules folder, we should explicitely list it for
+            // `removeDir` to delete it, but we need to delete it first, so we add it to inner delete list
+            if (prevNode && prevNode.children.has(NODE_MODULES))
+              innerDeleteList.add(ppath.join(curLocation, NODE_MODULES));
```

Comment:
Wouldn't it be enough to process the delete list in the reverse insertion order?

Reply:
No, since we run all delete operation in parallel